File: src/recursion/tree/BinaryTreeMaximumPathSum.java
package recursion.tree;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

import utility.TreeNode;

/**
Given a binary tree, find the maximum path sum.

For this problem, a path is defined as any sequence of nodes from some starting node to any node in the tree along the parent-child connections. 
The path does not need to go through the root.

For example:
Given the below binary tree,

       1
      / \
     2   3
Return 6.
 */



public class BinaryTreeMaximumPathSum
{	
	class ResultWrapper
	{
		public final int includeRootMax;
		public final int globalMax;
		public ResultWrapper( int includeRootMax, int globalMax )
		{
			this.includeRootMax = includeRootMax;
			this.globalMax = globalMax;
		}
	}
	
    public int maxPathSum( TreeNode root ) 
    {
    	if ( root == null )
    	{
    		return 0;
    	}
    	ResultWrapper result = maxPathSumRecurse( root );
    	return Math.max( result.globalMax, result.includeRootMax );
    }
    
    private ResultWrapper maxPathSumRecurse( TreeNode root )
    {
    	if ( root == null )
    	{
    		return new ResultWrapper( 0, Integer.MIN_VALUE );
    	}
    	if ( root.left == null 
    			&& root.right == null )
    	{
    		return new ResultWrapper( root.val, root.val );
    	}
    	
    	ResultWrapper leftChildResult = maxPathSumRecurse( root.left );
    	ResultWrapper rightChildResult = maxPathSumRecurse( root.right );
    	int includeRootMax = Math.max( root.val,
    										Math.max( leftChildResult.includeRootMax + root.val, rightChildResult.includeRootMax + root.val ) );
    	int globalMax = Math.max( root.val + Math.max( 0, leftChildResult.includeRootMax ) + Math.max( 0, rightChildResult.includeRootMax ), 
    										Math.max( leftChildResult.globalMax, rightChildResult.globalMax ) );

    	return new ResultWrapper( includeRootMax, globalMax );
    }
    
    @Test
    public void test()
    {
    	TreeNode node1 = new TreeNode( 1 );
    	TreeNode node2 = new TreeNode( 2 );
    	TreeNode node3 = new TreeNode( 3 );
    	node1.left = node2;
    	node1.right = node3;
    	assertEquals( 6, maxPathSum( node1 ) );    	    	
    }
    
    @Test
    public void edgeTest()
    {
    	TreeNode node4 = new TreeNode( -2 );
    	TreeNode node5 = new TreeNode( -1 );
    	node4.left = node5;
    	assertEquals( -1, maxPathSum( node4 ) );
    }
}

File: Python/BinaryTreeMaximumPathSum.py
# Definition for a binary tree node.
import unittest

class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class BinaryTreeMaximumPathSum(unittest.TestCase):

    def testPassThroughRoot(self):
        node1 = TreeNode(1)
        node2 = TreeNode(2)
        node3 = TreeNode(3)
        node1.left = node2
        node1.right = node3
        self.assertEqual(6, self.maxPathSum(node1))

    def testNotPassThroughRoot(self):
        node1 = TreeNode(-10)
        node2 = TreeNode(9)
        node3 = TreeNode(20)
        node4 = TreeNode(15)
        node5 = TreeNode(7)
        node1.left = node2
        node1.right = node3
        node3.left = node4
        node3.right = node5
        self.assertEqual(42, self.maxPathSum(node1))

    def testNegativeNode(self):
        node1 = TreeNode(-3)
        self.assertEqual(-3, self.maxPathSum(node1))

    def testSecond(self):
        node1 = TreeNode(2)
        node2 = TreeNode(-1)
        node1.left = node2
        self.assertEqual(2, self.maxPathSum(node1))

    def testBinaryTreeMaximumPathSum(self):
        node1 = TreeNode(5)
        node2 = TreeNode(4)
        node3 = TreeNode(8)
        node4 = TreeNode(11)
        node5 = TreeNode(13)
        node6 = TreeNode(4)
        node7 = TreeNode(7)
        node8 = TreeNode(2)
        node9 = TreeNode(1)
        node1.left = node2
        node1.right = node3
        node2.left = node4
        node4.left = node7
        node4.right = node8
        node3.left = node5
        node3.right = node6
        node6.right = node9
        self.assertEqual(48, self.maxPathSum(node1))

    def maxPathSum(self, root: TreeNode) -> int:
        globalMax, passThroughRootMax = self.maxPathSumRecursive(root)
        return globalMax

    def maxPathSumRecursive(self, root: TreeNode) -> (int, int):
        if root is None:
            return (float('-inf'), 0)

        if root.left is None and root.right is None:
            return root.val, root.val

        (leftGlobalMax, leftChildRootedMax) = self.maxPathSumRecursive(root.left)
        (rightGlobalMax, rightChildRootedMax) = self.maxPathSumRecursive(root.right)

        passThroughRootMax = root.val + max(max(leftChildRootedMax, 0), max(rightChildRootedMax, 0))
        globalMax = max(root.val + max(leftChildRootedMax, 0) + max(rightChildRootedMax, 0), leftGlobalMax, rightGlobalMax)

        return (globalMax, passThroughRootMax)

if __name__ == '__main__':
    unittest.main()